Dependency problem on OpenSSL version via Cryptography package.

It might happen that during the automatic installation of ``more.jwauth`` , e.g. via Buildout, the version of the OpenSSL library linked to the Cryptography package is not recent enough. When that happens, Cryptography disables some encryption algorithms, and this typically causes problems at run-time.

While one might think of how to fix this non-Python dependency in Buildout, I think that at least ``more.jwauth`` should detect this situation as early as possible and raise a helpful warning directing people to the instructions to manually install Cryptography on the appropriate OS.

When I ran the test-suite with a Cryptography package that was thus restricted, I obtained the following traceback, which could be useful as a source of inspiration on how to implement this safety check:

```
________________________________ test_encode_decode_with_ps384 _________________________________

    def test_encode_decode_with_ps384():
        identity_policy = JWTIdentityPolicy(
            algorithm='PS384',
            private_key=get_data('tests/keys/testkey_rsa'),
            public_key=get_data('tests/keys/testkey_rsa.pub')
        )
        claims_set = {
            'sub': 'user'
        }
>       token = identity_policy.encode_jwt(claims_set)

more/jwtauth/tests/test_jwtauth.py:122: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
more/jwtauth/main.py:229: in encode_jwt
    token = jwt.encode(claims_set, key, algorithm)
jwt/api_jwt.py:56: in encode
    json_payload, key, algorithm, headers, json_encoder
jwt/api_jws.py:98: in encode
    signature = alg_obj.sign(signing_input, key)
jwt/algorithms.py:268: in sign
    self.hash_alg()
cryptography/hazmat/backends/openssl/rsa.py:531: in signer
    return _RSASignatureContext(self._backend, self, padding, algorithm)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cryptography.hazmat.backends.openssl.rsa._RSASignatureContext object at 0x1042b0850>
backend = <cryptography.hazmat.backends.openssl.backend.Backend object at 0x103d17310>
private_key = <cryptography.hazmat.backends.openssl.rsa._RSAPrivateKey object at 0x1042b0750>
padding = <cryptography.hazmat.primitives.asymmetric.padding.PSS object at 0x1042b0790>
algorithm = <cryptography.hazmat.primitives.hashes.SHA384 object at 0x1042b0810>

    def __init__(self, backend, private_key, padding, algorithm):
        self._backend = backend
        self._private_key = private_key
    
        if not isinstance(padding, AsymmetricPadding):
            raise TypeError("Expected provider of AsymmetricPadding.")
    
        self._pkey_size = self._backend._lib.EVP_PKEY_size(
            self._private_key._evp_pkey
        )
        self._backend.openssl_assert(self._pkey_size > 0)
    
        if isinstance(padding, PKCS1v15):
            if self._backend._lib.Cryptography_HAS_PKEY_CTX:
                self._finalize_method = self._finalize_pkey_ctx
                self._padding_enum = self._backend._lib.RSA_PKCS1_PADDING
            else:
                self._finalize_method = self._finalize_pkcs1
        elif isinstance(padding, PSS):
            if not isinstance(padding._mgf, MGF1):
                raise UnsupportedAlgorithm(
                    "Only MGF1 is supported by this backend.",
                    _Reasons.UNSUPPORTED_MGF
                )
    
            # Size of key in bytes - 2 is the maximum
            # PSS signature length (salt length is checked later)
            if self._pkey_size - algorithm.digest_size - 2 < 0:
                raise ValueError("Digest too large for key size. Use a larger "
                                 "key.")
    
            if not self._backend._mgf1_hash_supported(padding._mgf._algorithm):
                raise UnsupportedAlgorithm(
                    "When OpenSSL is older than 1.0.1 then only SHA1 is "
                    "supported with MGF1.",
>                   _Reasons.UNSUPPORTED_HASH
                )
E               UnsupportedAlgorithm: When OpenSSL is older than 1.0.1 then only SHA1 is supported with MGF1.

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,10 +2,13 @@ language: python
 
 cache: pip
 env:
-    - TOXENV=py27
-    - TOXENV=py33
-    - TOXENV=py34
-    - TOXENV=pep8
+    - TOXENV=py27-crypto
+    - TOXENV=py33-crypto
+    - TOXENV=py34-crypto
+    - TOXENV=py27-nocrypto
+    - TOXENV=py33-nocrypto
+    - TOXENV=py34-nocrypto
+    - TOXENV=flake8
     - TOXENV=coverage
 before_install:
     - pip install --upgrade pip setuptools

--- a/README.rst
+++ b/README.rst
@@ -60,7 +60,7 @@ You can use pip for installing more.jwtauth:
 * ``pip install -U more.jwtauth`` - installing without cryptography
 
 Alternatively you can use buildout to install more.jwtauth.
-Remember to install cryptography if you need it.
+Remember to install cryptography manually if you need it.
 
 
 Usage
@@ -164,34 +164,34 @@ HS384
 HS512
    HMAC using SHA-512 hash algorithm
 
-ES256 [*]_
+ES256 [1]_
    ECDSA signature algorithm using SHA-256 hash algorithm
 
-ES384 [*]_
+ES384 [1]_
    ECDSA signature algorithm using SHA-384 hash algorithm
 
-ES512 [*]_
+ES512 [1]_
    ECDSA signature algorithm using SHA-512 hash algorithm
 
-PS256 [*]_
+PS256 [1]_
    RSASSA-PSS signature using SHA-256 and MGF1 padding with SHA-256
 
-PS384 [*]_
+PS384 [1]_
    RSASSA-PSS signature using SHA-384 and MGF1 padding with SHA-384
 
-PS512 [*]_
+PS512 [1]_
    RSASSA-PSS signature using SHA-512 and MGF1 padding with SHA-512
 
-RS256 [*]_
+RS256 [1]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-256 hash algorithm
 
-RS384 [*]_
+RS384 [1]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-384 hash algorithm
 
-RS512 [*]_
+RS512 [1]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-512 hash algorithm
 
-.. [*] The marked algorithms require this library to be installed
+.. [1] The marked algorithms require this library to be installed
        with its ``crypto`` dependencies::
 
          pip install -U more.jwtauth[crypto]

--- a/README.rst
+++ b/README.rst
@@ -164,32 +164,40 @@ HS384
 HS512
    HMAC using SHA-512 hash algorithm
 
-ES256
+ES256 [*]_
    ECDSA signature algorithm using SHA-256 hash algorithm
 
-ES384
+ES384 [*]_
    ECDSA signature algorithm using SHA-384 hash algorithm
 
-ES512
+ES512 [*]_
    ECDSA signature algorithm using SHA-512 hash algorithm
 
-PS256
+PS256 [*]_
    RSASSA-PSS signature using SHA-256 and MGF1 padding with SHA-256
 
-PS384
+PS384 [*]_
    RSASSA-PSS signature using SHA-384 and MGF1 padding with SHA-384
 
-PS512
+PS512 [*]_
    RSASSA-PSS signature using SHA-512 and MGF1 padding with SHA-512
 
-RS256
+RS256 [*]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-256 hash algorithm
 
-RS384
+RS384 [*]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-384 hash algorithm
 
-RS512
+RS512 [*]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-512 hash algorithm
+
+.. [*] The marked algorithms require this library to be installed
+       with its ``crypto`` dependencies::
+
+         pip install -U more.jwtauth[crypto]
+
+       See Installation for details. In case of problems be sure
+       to have read the note in the Requirements section.
 
 
 Settings

--- a/README.rst
+++ b/README.rst
@@ -196,8 +196,8 @@ RS512 [1]_
 
          pip install -U more.jwtauth[crypto]
 
-       See Installation for details. In case of problems be sure
-       to have read the note in the Requirements section.
+       See Installation_ for details. In case of problems be sure
+       to have read the note in the Requirements_ section.
 
 
 Settings

--- a/README.rst
+++ b/README.rst
@@ -191,7 +191,7 @@ RS384 [1]_
 RS512 [1]_
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-512 hash algorithm
 
-.. [1] The marked algorithms require this library to be installed
+.. [1] The marked algorithms require more.jwtauth to be installed
        with its ``crypto`` dependencies::
 
          pip install -U more.jwtauth[crypto]

--- a/README.rst
+++ b/README.rst
@@ -35,6 +35,33 @@ You can include all necessary information about the identity in the token so JWT
 can be used by a stateless service e.g. with external password validation.
 
 
+Requirements
+------------
+
+-  Python (2.7, 3.3, 3.4, 3.5)
+-  morepath (>= 0.14)
+-  PyJWT (1.4.0)
+-  optional: cryptography (1.3.1)
+
+.. Note::
+   If you want to use another algorithm than HMAC (HS*), you need to install
+   cryptography.
+   On some systems this can be a little tricky. Please follow the instructions
+   in https://cryptography.io/en/latest/installation and be sure to install all
+   dependencies as referenced.
+
+
+Install
+-------
+
+You can use pip for installing more.jwtauth:
+
+* ``pip install -U more.jwtauth[crypto]`` - installing with cryptography
+* ``pip install -U more.jwtauth`` - installing without cryptography
+
+Alternatively you can use buildout to install more.jwtauth.
+Remember to install cryptography if you need it.
+
 
 Usage
 -----
@@ -121,14 +148,6 @@ Advanced:
             leeway=10
         )
 
-Requirements
-------------
-
--  Python (2.7, 3.3, 3.4, 3.5)
--  morepath (>= 0.14)
--  PyJWT (1.4.0)
--  cryptography (1.3.1 - be sure to install all dependencies as referenced in https://cryptography.io/en/latest/installation)
-
 
 Algorithms
 ----------
@@ -171,6 +190,7 @@ RS384
 
 RS512
    RSASSA-PKCS1-v1_5 signature algorithm using SHA-512 hash algorithm
+
 
 Settings
 --------

--- a/README.rst
+++ b/README.rst
@@ -51,8 +51,8 @@ Requirements
    dependencies as referenced.
 
 
-Install
--------
+Installation
+------------
 
 You can use pip for installing more.jwtauth:
 

--- a/more/jwtauth/main.py
+++ b/more/jwtauth/main.py
@@ -36,11 +36,11 @@ In the later case the algorithm must be an EC*, PS* or RS* version.
 """
 
 
-import sys
 import datetime
-import jwt
+import sys
 
-from morepath import (Identity, NO_IDENTITY)
+import jwt
+from morepath import NO_IDENTITY, Identity
 
 PY3 = sys.version_info[0] == 3
 

--- a/more/jwtauth/tests/test_cryptography.py
+++ b/more/jwtauth/tests/test_cryptography.py
@@ -16,7 +16,7 @@ def relative(filepath):
 @pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
 class TestCryptographyAlgorithms:
 
-    def test_encode_decode_with_es256():
+    def test_encode_decode_with_es256(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='ES256',
             private_key_file=relative('keys/testkey_ec'),
@@ -30,7 +30,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_es384():
+    def test_encode_decode_with_es384(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='ES384',
             private_key_file=relative('keys/testkey_ec'),
@@ -44,7 +44,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_es512():
+    def test_encode_decode_with_es512(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='ES512',
             private_key_file=relative('keys/testkey_ec'),
@@ -58,7 +58,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_ps256():
+    def test_encode_decode_with_ps256(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='PS256',
             private_key_file=relative('keys/testkey_rsa'),
@@ -72,7 +72,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_ps384():
+    def test_encode_decode_with_ps384(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='PS384',
             private_key_file=relative('keys/testkey_rsa'),
@@ -86,7 +86,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_ps512():
+    def test_encode_decode_with_ps512(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='PS512',
             private_key_file=relative('keys/testkey_rsa'),
@@ -100,7 +100,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_rs256():
+    def test_encode_decode_with_rs256(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='RS256',
             private_key_file=relative('keys/testkey_rsa'),
@@ -114,7 +114,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_rs384():
+    def test_encode_decode_with_rs384(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='RS384',
             private_key_file=relative('keys/testkey_rsa'),
@@ -128,7 +128,7 @@ class TestCryptographyAlgorithms:
 
         assert claims_set_decoded == claims_set
 
-    def test_encode_decode_with_rs512():
+    def test_encode_decode_with_rs512(self):
         identity_policy = JWTIdentityPolicy(
             algorithm='RS512',
             private_key_file=relative('keys/testkey_rsa'),

--- a/more/jwtauth/tests/test_cryptography.py
+++ b/more/jwtauth/tests/test_cryptography.py
@@ -1,0 +1,143 @@
+from more.jwtauth import JWTIdentityPolicy
+import pytest
+
+try:
+    from cryptography.hazmat.primitives.asymmetric import ec  # noqa
+    has_crypto = True
+except ImportError:
+    has_crypto = False
+
+
+def relative(filepath):
+    import os
+    return os.path.join(os.path.dirname(__file__), filepath)
+
+
+@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
+class TestCryptographyAlgorithms:
+
+    def test_encode_decode_with_es256():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='ES256',
+            private_key_file=relative('keys/testkey_ec'),
+            public_key_file=relative('keys/testkey_ec.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_es384():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='ES384',
+            private_key_file=relative('keys/testkey_ec'),
+            public_key_file=relative('keys/testkey_ec.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_es512():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='ES512',
+            private_key_file=relative('keys/testkey_ec'),
+            public_key_file=relative('keys/testkey_ec.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_ps256():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='PS256',
+            private_key_file=relative('keys/testkey_rsa'),
+            public_key_file=relative('keys/testkey_rsa.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_ps384():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='PS384',
+            private_key_file=relative('keys/testkey_rsa'),
+            public_key_file=relative('keys/testkey_rsa.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_ps512():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='PS512',
+            private_key_file=relative('keys/testkey_rsa'),
+            public_key_file=relative('keys/testkey_rsa.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_rs256():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='RS256',
+            private_key_file=relative('keys/testkey_rsa'),
+            public_key_file=relative('keys/testkey_rsa.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_rs384():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='RS384',
+            private_key_file=relative('keys/testkey_rsa'),
+            public_key_file=relative('keys/testkey_rsa.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set
+
+    def test_encode_decode_with_rs512():
+        identity_policy = JWTIdentityPolicy(
+            algorithm='RS512',
+            private_key_file=relative('keys/testkey_rsa'),
+            public_key_file=relative('keys/testkey_rsa.pub')
+        )
+        claims_set = {
+            'sub': 'user'
+        }
+        token = identity_policy.encode_jwt(claims_set)
+        claims_set_decoded = identity_policy.decode_jwt(token)
+
+        assert claims_set_decoded == claims_set

--- a/more/jwtauth/tests/test_cryptography.py
+++ b/more/jwtauth/tests/test_cryptography.py
@@ -1,5 +1,6 @@
-from more.jwtauth import JWTIdentityPolicy
 import pytest
+
+from more.jwtauth import JWTIdentityPolicy
 
 try:
     from cryptography.hazmat.primitives.asymmetric import ec  # noqa

--- a/more/jwtauth/tests/test_jwtauth.py
+++ b/more/jwtauth/tests/test_jwtauth.py
@@ -1,13 +1,14 @@
 # -*- coding: utf-8 -*-
 import datetime
-import morepath
-from morepath import (Response, settings, Identity, NO_IDENTITY)
 
+import pytest
+
+import morepath
+from jwt import InvalidIssuerError
 from more.jwtauth import JWTIdentityPolicy
+from morepath import NO_IDENTITY, Identity, Response, settings
 from webob.exc import HTTPProxyAuthenticationRequired
 from webtest import TestApp as Client
-from jwt import InvalidIssuerError
-import pytest
 
 
 def setup_module(module):

--- a/more/jwtauth/tests/test_jwtauth.py
+++ b/more/jwtauth/tests/test_jwtauth.py
@@ -9,6 +9,12 @@ from webtest import TestApp as Client
 from jwt import InvalidIssuerError
 import pytest
 
+try:
+    from cryptography.hazmat.primitives.asymmetric import ec  # noqa
+    has_crypto = True
+except ImportError:
+    has_crypto = False
+
 
 def setup_module(module):
     morepath.disable_implicit()
@@ -91,6 +97,7 @@ def test_encode_decode_with_issuer():
     assert claims_set_decoded == claims_set
 
 
+@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
 def test_encode_decode_with_es256():
     identity_policy = JWTIdentityPolicy(
         algorithm='ES256',
@@ -106,6 +113,7 @@ def test_encode_decode_with_es256():
     assert claims_set_decoded == claims_set
 
 
+@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
 def test_encode_decode_with_ps384():
     identity_policy = JWTIdentityPolicy(
         algorithm='PS384',
@@ -121,6 +129,7 @@ def test_encode_decode_with_ps384():
     assert claims_set_decoded == claims_set
 
 
+@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
 def test_encode_decode_with_rs512():
     identity_policy = JWTIdentityPolicy(
         algorithm='RS512',

--- a/more/jwtauth/tests/test_jwtauth.py
+++ b/more/jwtauth/tests/test_jwtauth.py
@@ -9,20 +9,9 @@ from webtest import TestApp as Client
 from jwt import InvalidIssuerError
 import pytest
 
-try:
-    from cryptography.hazmat.primitives.asymmetric import ec  # noqa
-    has_crypto = True
-except ImportError:
-    has_crypto = False
-
 
 def setup_module(module):
     morepath.disable_implicit()
-
-
-def relative(filepath):
-    import os
-    return os.path.join(os.path.dirname(__file__), filepath)
 
 
 def test_jwt_custom_settings():
@@ -91,54 +80,6 @@ def test_encode_decode_with_issuer():
         'iss': 'Issuer_App'
     }
     claims_set = identity_policy.create_claims_set(userid, extra_claims)
-    token = identity_policy.encode_jwt(claims_set)
-    claims_set_decoded = identity_policy.decode_jwt(token)
-
-    assert claims_set_decoded == claims_set
-
-
-@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
-def test_encode_decode_with_es256():
-    identity_policy = JWTIdentityPolicy(
-        algorithm='ES256',
-        private_key_file=relative('keys/testkey_ec'),
-        public_key_file=relative('keys/testkey_ec.pub')
-    )
-    claims_set = {
-        'sub': 'user'
-    }
-    token = identity_policy.encode_jwt(claims_set)
-    claims_set_decoded = identity_policy.decode_jwt(token)
-
-    assert claims_set_decoded == claims_set
-
-
-@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
-def test_encode_decode_with_ps384():
-    identity_policy = JWTIdentityPolicy(
-        algorithm='PS384',
-        private_key_file=relative('keys/testkey_rsa'),
-        public_key_file=relative('keys/testkey_rsa.pub')
-    )
-    claims_set = {
-        'sub': 'user'
-    }
-    token = identity_policy.encode_jwt(claims_set)
-    claims_set_decoded = identity_policy.decode_jwt(token)
-
-    assert claims_set_decoded == claims_set
-
-
-@pytest.mark.skipif(not has_crypto, reason='Not supported without cryptography library')
-def test_encode_decode_with_rs512():
-    identity_policy = JWTIdentityPolicy(
-        algorithm='RS512',
-        private_key_file=relative('keys/testkey_rsa'),
-        public_key_file=relative('keys/testkey_rsa.pub')
-    )
-    claims_set = {
-        'sub': 'user'
-    }
     token = identity_policy.encode_jwt(claims_set)
     claims_set_decoded = identity_policy.decode_jwt(token)
 

--- a/setup.py
+++ b/setup.py
@@ -44,7 +44,7 @@ setup(
         'PyJWT == 1.4.0',
     ],
     extras_require=dict(
-        cryptography=['cryptography == 1.3.1'],
+        crypto=['cryptography == 1.3.1'],
         test=['pytest >= 2.9.1',
               'pytest-cov',
               'WebTest'],

--- a/setup.py
+++ b/setup.py
@@ -41,9 +41,9 @@ setup(
         'setuptools',
         'morepath >= 0.14',
         'PyJWT == 1.4.0',
-        'cryptography == 1.3.1'
     ],
     extras_require=dict(
+        cryptography=['cryptography == 1.3.1'],
         test=['pytest >= 2.9.1',
               'pytest-cov',
               'WebTest'],

--- a/setup.py
+++ b/setup.py
@@ -1,5 +1,6 @@
 import io
-from setuptools import setup, find_packages
+
+from setuptools import find_packages, setup
 
 long_description = (
     io.open('README.rst', encoding='utf-8').read() + '\n\n' +

--- a/tox.ini
+++ b/tox.ini
@@ -2,7 +2,6 @@
 envlist = py{27,33,34}-crypto, py{27,33,34}-nocrypto, flake8, coverage
 
 [testenv]
-deps = -e{toxinidir}[test]
 commands = py.test -v {posargs}
 deps =
     -e{toxinidir}[test]

--- a/tox.ini
+++ b/tox.ini
@@ -1,18 +1,26 @@
 [tox]
-envlist = py27,py33,py34,pep8,coverage
+envlist = py{27,33,34}-crypto, py{27,33,34}-nocrypto, flake8, coverage
 
 [testenv]
 deps = -e{toxinidir}[test]
 commands = py.test -v {posargs}
+deps =
+    -e{toxinidir}[test]
+    crypto: cryptography
 
-[testenv:pep8]
+[testenv:flake8]
 basepython = python2
-deps = flake8
 commands = flake8
+deps =
+    flake8
+    flake8-import-order
+    pep8-naming
 
 [testenv:coverage]
 basepython = python3
-deps = {[testenv]deps}
+deps =
+    {[testenv]deps}
+    cryptography
 commands =
     coverage run --source more.jwtauth -m py.test {posargs}
     coverage report -m --fail-under=100

--- a/tox.ini
+++ b/tox.ini
@@ -12,7 +12,6 @@ basepython = python2
 commands = flake8
 deps =
     flake8
-    flake8-import-order
     pep8-naming
 
 [testenv:coverage]

--- a/tox.ini
+++ b/tox.ini
@@ -17,7 +17,7 @@ deps =
 [testenv:coverage]
 basepython = python3
 deps =
-    {[testenv]deps}
+    -e{toxinidir}[test]
     cryptography
 commands =
     coverage run --source more.jwtauth -m py.test {posargs}

